internal/primitives/state-machine/overlayedchanges/overlayed_changes.go

```diff
@@ -65,3 +73,607 @@ type IndexOperationRenew struct {
+		for key, value := range child.Changes() {
+			deltas = append(deltas, backend.Delta{Key: []byte(key), Value: value.Value()})
+		}
+
+		childDeltas = append(childDeltas, backend.ChildDelta{
+			ChildInfo: child.ChildInfo,
+			Deltas:    deltas,
+		})
+	}
+
+	root, tx := b.FullStorageRoot(delta, childDeltas, stateVersion)
+	oc.storageTransactionCache = &StorageTransactionCache[H, Hasher]{
+		transaction:            tx,
+		transactionStorageRoot: root,
+	}
+
+	return root, false
+}
+
+func (oc *OverlayedChanges[H, Hasher]) ChildStorageRoot(
+	childInfo storage.ChildInfo,
+	b backend.Backend[H, Hasher],
+	stateVersion storage.StateVersion,
+) (H, bool, error) {
+	storageKey := StorageKey(childInfo.StorageKey())
+	prefixedStorageKey := childInfo.PrefixedStorageKey()
+
+	var root H
+
+	if oc.storageTransactionCache != nil {
+		value, has := oc.Storage(string(prefixedStorageKey))
+		if !has {
+			backendValue, err := b.Storage(prefixedStorageKey)
+			if err != nil {
+				return trie.EmptyChildTrieRoot[H, Hasher](), false, err
+			}
+
+			if backendValue == nil || scale.Unmarshal(backendValue, &value) != nil {
+				root = trie.EmptyChildTrieRoot[H, Hasher]()
+			}
+		} else {
+			hasher := *new(Hasher)
+			root = hasher.NewHash(value)
+		}
+
+		return root, true, nil
+	}
+
+	var calculatedRoot H
+	var isEmpty bool
+
+	changes, info := oc.ChildChanges(storageKey)
+	if changes == nil || info == nil {
+		root = trie.EmptyChildTrieRoot[H, Hasher]()
+	} else {
+		delta := make([]backend.Delta, 0)
+		for k, v := range changes {
+			delta = append(delta, backend.Delta{Key: []byte(k), Value: v.Value()})
+		}
+		calculatedRoot, isEmpty, _ = b.ChildStorageRoot(info, delta, stateVersion)
+	}
+
+	if calculatedRoot != trie.EmptyChildTrieRoot[H, Hasher]() {
+		if isEmpty {
+			oc.SetStorage(StorageKey(prefixedStorageKey), nil)
+		} else {
+			oc.SetStorage(StorageKey(prefixedStorageKey), calculatedRoot.MustMarshalSCALE())
+		}
+		oc.markDirty()
+		root = calculatedRoot
+	} else {
+		// empty overlay
+		backendValue, err := b.Storage(prefixedStorageKey)
+		if err != nil {
+			return trie.EmptyChildTrieRoot[H, Hasher](), false, err
+		}
+
+		err = scale.Unmarshal(backendValue, &root)
+		if err != nil {
+			root = trie.EmptyChildTrieRoot[H, Hasher]()
+		}
+	}
+
+	return root, false, nil
+}
+
+func (oc *OverlayedChanges[H, Hasher]) IterAfter(key StorageKey) iter.Seq2[StorageKey,
+	*OverlayedStorageEntry] {
+	return oc.top.changesAfter(key)
+}
+
+func (oc *OverlayedChanges[H, Hasher]) ChildIterAfter(
+	storageKey StorageKey,
+	key StorageKey,
+) iter.Seq2[StorageKey, *overlayedValue] {
+	entry, has := oc.children[string(storageKey)]
+	if !has {
+		return nil
+	}
+
+	return entry.overlayedChangeSet.changesAfter(key)
+}
+
+func (oc *OverlayedChanges[H, Hasher]) SetOffchainStorage(key []byte, value []byte) {
+	if value == nil {
+		oc.offchain.Remove(offchain.StoragePrefix, key)
+	} else {
+		oc.offchain.Set(offchain.StoragePrefix, key, value)
+	}
+}
```

Comment:
There's a method missing vs the substrate impl `add_transaction_index`.  Will we need it?